components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java

```diff
@@ -125,6 +131,55 @@ public OAuth2AuthorizeRespDTO authorize(OAuth2AuthorizeReqDTO oAuth2AuthorizeReq
         }
     }
 
+    /**
+     * Process the authorization request and issue an authorization code or access token depending
+     * on the Response Type available in the request.
+     *
+     * @param authzReqMsgCtx authzReqMsgCtx.
+     * @return <code>OAuth2AuthorizeRespDTO</code> instance containing the access token/authorization code
+     * or an error code.
+     */
+    public OAuth2AuthorizeRespDTO authorize(OAuthAuthzReqMessageContext authzReqMsgCtx) {
+        OAuth2AuthorizeReqDTO authzReqDTO = authzReqMsgCtx.getAuthorizationReqDTO();
+        if (log.isDebugEnabled()) {
+            log.debug("Authorization Request received for user : " + authzReqDTO.getUser() +
+                    ", Client ID : " + authzReqDTO.getConsumerKey() +
+                    ", Authorization Response Type : " + authzReqDTO.getResponseType() +
+                    ", Requested callback URI : " + authzReqDTO.getCallbackUrl() +
+                    ", Requested Scopes : " + OAuth2Util.buildScopeString(authzReqMsgCtx.getApprovedScope()) +
+                    ", Approved Scopes : " + OAuth2Util.buildScopeString(
+                    authzReqDTO.getScopes()));
+        }
+        try {
+            AuthorizationHandlerManager authzHandlerManager = AuthorizationHandlerManager.getInstance();
+            return authzHandlerManager.handleAuthorization(authzReqMsgCtx);
+        } catch (Exception e) {
+            LoggerUtils.triggerDiagnosticLogEvent(OAuthConstants.LogConstants.OAUTH_INBOUND_SERVICE, null,
+                    OAuthConstants.LogConstants.FAILED, "Error occurred when processing the authorization request.",
+                    "authorize-client", null);
+            log.error("Error occurred when processing the authorization request. Returning an error back to client.",
+                    e);
+            OAuth2AuthorizeRespDTO authorizeRespDTO = new OAuth2AuthorizeRespDTO();
+            authorizeRespDTO.setErrorCode(OAuth2ErrorCodes.SERVER_ERROR);
+            authorizeRespDTO.setErrorMsg("Error occurred when processing the authorization " +
+                    "request. Returning an error back to client.");
+            authorizeRespDTO.setCallbackURI(authzReqDTO.getCallbackUrl());
+            return authorizeRespDTO;
+        }
+    }
+
+    /**
+     * Handle authorization request (validate requested scopes) before the consent page.
+     *
+     * @param authzReqDTO OAuth2AuthorizeReqDTO
+     * @return OAuthAuthzReqMessageContext
+     */
+    public OAuthAuthzReqMessageContext validateScopesBeforeConsent(OAuth2AuthorizeReqDTO authzReqDTO)
```

Comment:
// Let's add a comment in this method as to why we have to return this context instead of a response object